test_runner: add validation to test `options` argument in harness.js

GhoulKingR opened the pull request:
<!--
Before submitting a pull request, please read:

- the CONTRIBUTING guide at https://github.com/nodejs/node/blob/HEAD/CONTRIBUTING.md
- the commit message formatting guidelines at
  https://github.com/nodejs/node/blob/HEAD/doc/contributing/pull-requests.md#commit-message-guidelines

For code changes:
1. Include tests for any bug fixes or new features.
2. Update documentation if relevant.
3. Ensure that `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes.

If you believe this PR should be highlighted in the Node.js CHANGELOG
please add the `notable-change` label.

Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
-->

test_runner: add validation to test `options` argument in harness.js

In the `runInParentContext` function, I created a new function called `validateTestOptions` to check for invalid options and throw an error if it encounters one. I then made the function call in the `run` function inside `runInParentContext`. The `run` function checks if it is working with a `Test` instance before checking.

I also added a new failure type to `test_runner/test.js` called `kInvalidTestOptions`.

Fixes: https://github.com/nodejs/node/issues/51161

ljharb commented on lib/internal/test_runner/harness.js:
```diff
@@ -217,6 +218,18 @@ async function startSubtest(subtest) {
+  function validateTestOptions(options) {
+    const validOptions = ['concurrency', 'only', 'signal', 'skip', 'todo', 'timeout'];
+    for (const key of ObjectKeys(options)) {
+      if (validOptions.findIndex((v) => v === key) === -1) {
```
```suggestion
      if (ArrayPrototypeFindIndex(validOptions, (v) => v === key) === -1) {
```

GhoulKingR replied:
added 👍

ljharb commented on lib/internal/test_runner/harness.js:
```diff
@@ -217,6 +218,18 @@ async function startSubtest(subtest) {
 function runInParentContext(Factory) {
+  function validateTestOptions(options) {
+    const validOptions = ['concurrency', 'only', 'signal', 'skip', 'todo', 'timeout'];
+    for (const key of ObjectKeys(options)) {
```
this should be ArrayPrototypeForEach instead of for-of

GhoulKingR replied:
I've converted it

ljharb commented on lib/internal/test_runner/harness.js:
```diff
@@ -217,6 +219,18 @@ async function startSubtest(subtest) {
 
 function runInParentContext(Factory) {
+  function validateTestOptions(options) {
+    const validOptions = ['concurrency', 'only', 'signal', 'skip', 'todo', 'timeout'];
```
this should also be defined at module level so you don’t need a new array every time - and if it’s a SafeSet instead of an array it’ll be faster and simpler to do the lookup - you won’t need primordials.

GhoulKingR replied:
Did I do it correctly?

hulkish replied:
shouldn't this actually validate against a schema instead of just looking for accepted options?

for example, if I do `concurrency: 'foo'` or `concurrency: Infinity` it should throw

ljharb replied:
Good call; yes, it should check the types of the input too.

GhoulKingR replied:
That check is already covered as far as I saw. The only issue I saw was that it didn't check for what keys were passed to the options.

But if I am wrong, do let me know so that I can look into it.

GhoulKingR replied:
> Good call; yes, it should check the types of the input too.

Alright then, I'll work on it

hulkish replied:
In addition to validating that only accepted options are provided (which you have so far) - it should also perform strict type checking as per defined in the [docs](https://nodejs.org/dist/latest-v21.x/docs/api/test.html#testname-options-fn):

```
concurrency: <number> | <boolean>
only: <boolean>
signal: <AbortSignal>
skip: <boolean>
todo: <boolean> | <number>
```

In addition to this, I would argue that since `Infinity` is technically a valid number value... it should probably throw an error when you try `concurrency: Infinity`

ljharb replied:
Not my call for that question, but I'd suggest the latter, since that matches "max available".

hulkish replied:
So actually handling of a `Infinity` value is already treated as a not-accepted value for this via https://github.com/nodejs/node/blob/main/lib/internal/validators.js#L142

GhoulKingR replied:
I don't think I covered the possibility of infinity. But I did a basic type check for all the keys. Is that going to be okay? Do I need to cover it too?

ljharb commented:
Do note this is a breaking change for anyone passing in an unknown option.

ljharb commented on lib/internal/test_runner/harness.js:
```diff
@@ -217,6 +220,17 @@ async function startSubtest(subtest) {
 
 function runInParentContext(Factory) {
+  function validateTestOptions(options) {
+    ArrayPrototypeForEach(ObjectKeys(options), (key) => {
```
wouldn’t this still allow me to pass `{ __proto__: whatever }` and it’d ignore those keys, but a Get later would still pull out their values? This might need to use for-in.

GhoulKingR replied:
Hi, I'm still trying to fully understand what you mean. `ArrayPrototypeForEach` already skips the `Object.prototype` values. Do you want me to make it also check those values too?

ljharb replied:
what i mean is, `{ a, __proto__: { b } }` would be an options object where `options.a` and `options.b` would both work, but using ObjectKeys only would grab the `a`.

GhoulKingR replied:
So you want the validator to use a for-in loop that skips the `Object.prototype` keys?

ljharb replied:
I'm not really sure. but if you really want to check for extra keys, you'd cover anything that's inherited except for those on Object.prototype.

GhoulKingR replied:
Alright then, I have an idea of how to do that.

GhoulKingR replied:
I've added the for-in loop that skips every `__proto__` values

hulkish commented:
> Do note this is a breaking change for anyone passing in an unknown option.

true, but should be safe for landing in `21.x` since it's not LTS right?

ljharb commented:
Arguably yes for an experimental feature, but "not LTS" isn't license to break people indiscriminately.

atlowChemi commented on lib/internal/test_runner/harness.js:
```diff
@@ -217,6 +220,17 @@ async function startSubtest(subtest) {
 }
 
 function runInParentContext(Factory) {
+  function validateTestOptions(options) {
```
Any specific reason this function should be nested? (and not declared on top level)

GhoulKingR replied:
It was my initial thought to declare it nested. But seeing it now, it would be better declared at the top level. I added that change to my recent commit

atlowChemi reviewed (commented):
Not sure I see what the advantage of this change is.
Why can't one pass an object with additional values and still execute the tests?

ljharb commented:
It makes it easier to add options, because you can be sure nobody’s passing them already.

hulkish commented:
> Not sure I see what the advantage of this change is. Why can't one pass an object with additional values and still execute the tests?

I would argue that it in a test environment - it is in a test runner's best interest to set an expectation of correctness. Aka, when you have incorrect code in a test - it should throw an error

hulkish commented:
Hmm, so actually it seems that strict type checking is already being performed for accepted options: https://github.com/nodejs/node/blob/main/lib/internal/test_runner/test.js#L263 - So maybe just checking for accepted options is what's missing

hulkish commented:
@GhoulKingR see https://github.com/nodejs/node/pull/51197#issuecomment-1859240469

GhoulKingR commented:
@hulkish, so just to do validation for `todo`, `only` and `skip`?

hulkish commented:
> @hulkish, so just to do validation for `todo`, `only` and `skip`?

yea, and probably should do it where the other options are being validated: https://github.com/nodejs/node/blob/main/lib/internal/test_runner/test.js#L208

GhoulKingR commented:
I've added the validation for `todo`, `only`, and `skip`. I also moved the validations to `test.js`.

I tried moving the checks for invalid keys to the `Test` class earlier, but it didn't work well. So I left that validation in `harness.js`

ljharb commented on lib/internal/test_runner/harness.js:
```diff
@@ -216,6 +229,17 @@ async function startSubtest(subtest) {
 
+function validateTestOptions(options) {
+  for (const key in options) {
+    if (!(key in ObjectGetPrototypeOf(options)) && !validOptions.has(key)) {
```
```suggestion
    if (!(key in Object.prototype) && !validOptions.has(key)) {
```

GhoulKingR replied:
The linter tells me to use ObjectPrototype from premodials. Is that okay?

ljharb replied:
yep, i wasn't sure if that was a thing - definitely use that.

GhoulKingR replied:
Done ✅

aduh95 reviewed (commented):
I'm -0.5 on the idea, it's a very unusual pattern in JS to do such thing, and the fact that you have to special case `Object.prototype` is a good illustration why. It would probably make it harder to support different versions of Node.js.

cjihrig commented:
I agree with @aduh95 and others here, plus this seems to have stalled out. I'll go ahead and close this out. Thanks for the PR though.
